### PR TITLE
Convert `reduce_dim_naive` kernel to use the `#[cube]` derive macro

### PR DESCRIPTION
**Pull request by cBournhonesque**
## Pull Request Template

Following a discussion with @nathanielsimard on discord, I wanted to try to convert the reduce kernel to use the cube derive macro.
(originally because I wanted to know how to set the `CubeCount` for a kernel).

The new macro is pretty ergonomic, it's actually crazy that it also works for `impl Trait`.
I have a couple remaining issues though:

### Checklist

- [x] Implement `CubeType` for a tuple of `CubeType`s: https://github.com/tracel-ai/cubecl/issues/48
- [ ] How do I get the max value/min value of a `Numeric`? Do I need to add a trait for it?
- [x] Need to add the computation of the `CubeCount`
- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Testing

Unit tests

### CODE REVIEW

File: crates/burn-jit/src/kernel/reduce/naive/shader.rs [hunk 1]
```diff
@@ -1,148 +1,42 @@
...
-        let item_output = EO::cube_elem().into();
+#[cube(launch)]
+pub(crate) fn naive_reduce_dim_compute_shader<RD: ReduceDimNaive<EI>, EI: Numeric, EO: Numeric>(
+    input: &Tensor<EI>,
+    output: &mut Tensor<EO>,
+    dim: UInt,
+) {
+    if ABSOLUTE_POS >= output.len() {
+        return;
+    }
```
Comment (nathanielsimard) [inline crates/burn-jit/src/kernel/reduce/naive/shader.rs:15]: If you are doing bound checks, you can change `#[cube(launch)]` to `#[cube(launch_unchecked)`] and call `launch_unchecked` for improved performance.

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Some changes to be done, but that's crazy how flexible cube is, I never though we would be able to migrate the reduce trait to cube with that amount of ease.

File: crates/burn-jit/src/kernel/reduce/naive/base.rs [hunk 2]
```diff
@@ -1,32 +1,39 @@
...
+        accumulator: <Self::Accumulator as cubecl::frontend::CubeType>::ExpandType,
+        current_value: <EI as cubecl::frontend::CubeType>::ExpandType,
+        _i: <UInt as cubecl::frontend::CubeType>::ExpandType,
+    );
+
+    fn __expand_assign_naive<EO: Numeric>(
+        context: &mut cubecl::frontend::CubeContext,
+        output: <Tensor<EO> as cubecl::frontend::CubeType>::ExpandType,
+        accumulator: <Self::Accumulator as cubecl::frontend::CubeType>::ExpandType,
+        _shape_reduce_dim: <UInt as cubecl::frontend::CubeType>::ExpandType,
```
[the remark this answers is not in the document]
Comment (cBournhonesque) [reply crates/burn-jit/src/kernel/reduce/naive/base.rs:37]: oh nice!

File: crates/burn-jit/src/kernel/reduce/naive/shader.rs (hunk 1 above)
Comment (cBournhonesque) [reply crates/burn-jit/src/kernel/reduce/naive/shader.rs:15]: done!

**Comment by cBournhonesque** [comment]
> Some changes to be done, but that's crazy how flexible cube is, I never though we would be able to migrate the reduce trait to cube with that amount of ease.

Yes, it's crazy that the code just works like normal rust but is still handled correctly! Very nice that it works across traits as well

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**